Location creation

beamalsky opened the pull request.

beamalsky commented:
Testing notes:
- I'm able to successfully delete locations, but not add
- Some relation searches work, but some throw this error:
<img width="986" alt="screen shot 2018-09-19 at 10 09 31 am" src="https://user-images.githubusercontent.com/1094243/45762539-2f9c8000-bbf4-11e8-8688-324b05d7efca.png">

- Location detail pages still need a map
- For node searches, the map and results table can collide: 

<img width="1034" alt="screen shot 2018-09-19 at 10 02 14 am" src="https://user-images.githubusercontent.com/1094243/45762068-2ced5b00-bbf3-11e8-8539-e083b14a4c68.png">

evz commented:
@beamalsky Let's get together on this tomorrow and I can point you in the direction of where to start working on fixes for these things.

beamalsky commented:
Fixes from comment above:
1. I needed to install gdal! (`brew install gdal` on OSX)
2. Added an exception for this error, which was caused by finding no results for a relation search. Also added a results count (eg, "Results (0)") to indicate empty searches.
3. Fixed column sizing + it works on mobile now and hides map for empty results.
4. Fixed map and result overlap.

Other things to do:
- [ ] Add ID numbers to map pop ups
- [ ] Get country codes working on location detail pages
- [ ] Consider adding more info to detail pages
- [ ] Add confirmation message for newly added locations
- [ ] Add loading animation to `Go!` button

beamalsky commented:
This app now searches for a location by OSM type, country, and name, and allow users to add and delete locations. I've switched to a different Overpass API instance (https://overpass.kumi.systems/), which means we no longer have a rate limit, but these queries are still slow. 

@evz, could you pull this down and click around when you have a chance?

Right now searches are within a set country, case insensitive, and use regex to search by both default and english name and for all nodes/relations starting with the search term. This could possibly be sped up by dropping some of those qualifiers.

beamalsky commented:
The design of this now better matches up with preexisting styles, and the locations listing page is searchable, filterable, and paginated!

The create page now asks you to put in a type (node, way, or relation) and ID number to pinpoint an OSM object—in starting to take that former search filter out, I realized that that IDs are not unique between those 3 types (for example, see https://www.openstreetmap.org/node/387198407 and https://www.openstreetmap.org/way/387198407).

This is getting close to ready to integrate with the rest of the site!

beamalsky commented:
One thing to note is that these changes mean we no longer have country codes associated with location objects. Do we need those/want to figure out a way to add them manually here?

evz commented on sfm_pc/settings.py:
```diff
@@ -239,49 +240,49 @@
-        'relation_id': '192787',
-        'country_code': 'ng',
-    },
+    # {
```
We need to make sure all this stuff is uncommented.

evz commented on templates/location/list.html:
```diff
@@ -0,0 +1,214 @@
+{% endblock %}
+{% block extra_js %}
+<script src="{% static "js/select2.min.js" %}"></script>
+<script src="{% static "js/bootstrap-datepicker.min.js" %}"></script>
```
A lot of this javascript is probably unnecessary. Actually, most of it.

evz commented:
Things are looking pretty good. However, it looks like somehow the geometry is not getting saved properly: 

![screenshot_2018-10-03 whowasincommand](https://user-images.githubusercontent.com/551491/46419467-98efb900-c6f3-11e8-98f8-a44b7d553334.png)

When I look in the table, the geometry looks like: 

```
sfm=# select st_asgeojson(geometry) from location_location where id = '305602522';
                             st_asgeojson                              
-----------------------------------------------------------------------
 {"type":"Point","coordinates":[-0.000996960720977,0.000168933807231]}
```

But it looks like the form has the correct value (which is weird):

![geom](https://user-images.githubusercontent.com/551491/46419669-00a60400-c6f4-11e8-98b1-bd76c8df539e.png)

So, let's get together on this and see if we can figure this out.

beamalsky commented:
This app is currently grabbing relations from Wambacher, which means that they must have an admin level to be included. Building relations directly from the Overpass API is doable but takes much longer, risking a timeout—@tlongers, do you have a sense of how often you'll be needing relations that don't have admin levels?

tlongers commented:
@beamalsky Tha
> do you have a sense of how often you'll be needing relations that don't have admin levels?
Thanks for the question - I've never considered this before. If we ran the IDs of the  Relations we already use through a quick check on OSM we could find out the exact answer to that. I suspect that most of the relations we use have been assigned an admin level, but a few may not have. If we give the relation ID and type, does Overpass still take an aeon to grab it?

tlongers commented:
@beamalsky @evz  I did some analysis on your question about whether or not the OSM relations we use will always be assigned an admin level in OSM, and hence be downloadable from the Wambacher boundary service. 

In our Nigeria data, we have used 206 distinct relations in the `Site:admin_level_1_geonames_id` and `Area of Operations GeonameID`. Of these relations there are only two that do not have an admin level assigned to them (`8659823` and `8659834`). BUT ... these two are actually errors; corrected, they also have admin levels assigned to them.  So, at least in Nigeria we only use relations with admin_level values.

I ran the analysis on the Mexico data, and the situation is the same. Out of 1670 distinct relations, six do not have a admin level, but they also appear to be errors. 

At least from the data we have created, it seems the relations we use always have an admin level attached to them. The problem I have is generalising out from this to say "we will never use a relation that does not have an admin level assigned by OSM".

evz commented:
@tlongers I'm going to go ahead and merge this so you guys can start taking a look at it. It's not wired up to the other parts of the app yet (meaning the locations that you add are just kind in their own universe for the moment) but you can at least take a look at how in functions and perhaps give some feedback on that. I'll let you know when it's deployed on the staging site.
